### PR TITLE
Fixes for execution of TRT, TRTR, ED, EDMK running in Amode 64

### LINKED ISSUE #596: TRT in Amode64 does not clear high word of R1
Reported by Bob Schulien, email dd 2025-01-16, 10:08

The TRT instruction does not clear the high half of R1 if the CPU is in AMODE 64.
      LGHI   R1,-1
     TRT     FIELD,TRT_TABLE
     IF (NZ)
          R1 high half is still high-values/not a 64 bit address.
    ENDIF
Workaround is to do a LGHI R1,0 before the TRT.

### PR DESCRIPTION
**Pull request by jyganci**
Fixes #596. Issue #596 reported that the TRT instruction did not correctly set 64-bit register 1 (GR1) when issued in 64-bit addressing mode. Research unconvered two more instructions -- TRTR and EDMK -- that have the same problem. Testing a preliminary fix uncovered additional problems with the ED and EDMK instructions. Finally, running the Groovy regression test for the mfacc subdirectory uncovered a bug in two of the test programs -- P9MM1.MLC and P9MM2.MLC -- where R1 was not set correctly prior to an EDMK instruction. Previous pz390.java EDMK code did not properly handle the invalid R1 value, resulting in the two programs appearing to run correctly.

All of these issues/problems are being fixed by this pull request.

### CODE REVIEW

File: rt/mlc/IS596.MLC [hunk 1]
```diff
@@ -0,0 +1,4364 @@
...
+*
+***********************************************************************
+*        Strings used in the TRT and TRTR tests
+***********************************************************************
+         DS    0D
+StrNone  DC    C'abcd'
+StrMid   DC    C'a  d'
+StrEnd   DC    C'abc '
+StrBeg   DC    C' bcd'
+         DC    16X'00'         Needed to avoid S0C5 due to pz390 bug
```
Comment (abekornelis) [inline rt/mlc/IS596.MLC]: Could you please create an issue to describe this bug?

File: rt/mlc/IS596.MLC [hunk 2]
```diff
@@ -0,0 +1,4364 @@
...
+*
+TTRT_TRT TRT   0(*-*,6),0(8)       First match using translate table
+*
+          DS    0D
+***********************************************************************
+*         TRT tests; TTRTDS DSECT overlays a test
+***********************************************************************
+*
+TRTTests  DC    A(TRTT1,TTRTDSLN,TRTTn,0)  A(1st),len 1,A(last),unused
+*
```
Comment (abekornelis) [inline rt/mlc/IS596.MLC:477]: There is no test using StrBeg - I'm sure you did that on purpose, but could you please add some comment to explain why this edge case is irrelevant?

File: rt/mlc/IS596.MLC [hunk 3]
```diff
@@ -0,0 +1,4364 @@
...
+         LTORG ,
+*
+TTRTR_TRTR TRTR 0(*-*,9),0(8)      Reverse 1st match via translate tbl
+*
+          DS    0D
+***********************************************************************
+*         TRTR tests; TTRTDS DSECT overlays a test
+***********************************************************************
+*
+TRTRTests DC   A(TRTRT1,TTRTDSLN,TRTRTn,0) A(1st),len 1,A(last),unused
```
Comment (abekornelis) [inline rt/mlc/IS596.MLC:859]: Test with StrEnd is missing, please explain why?

File: rt/mlc/IS596.MLC [hunk 4]
```diff
@@ -0,0 +1,4364 @@
...
+***********************************************************************
+*         Put new tests above this line
+***********************************************************************
+TRTRTn    EQU  *-TTRTDSLN          Last test
+*
+          DS    0D
+***********************************************************************
+*         TRTR tests; 31-bit only; preset, expected GR1 bit 32 = 0
+***********************************************************************
+*
```
Comment (abekornelis) [inline rt/mlc/IS596.MLC:911]: Tests for StrEnd and StrNone are missing, please explain why?

**Review by abekornelis — COMMENTED** [review]
@jyganci No remarks - I like your table-driven approach to test case handling.
I did add a few questions, though. I'll be happy to approve, whatever your replies may be.

File: rt/mlc/IS596.MLC (hunk 1 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC]: The pz390.java code is sloppy about going beyond of memory. It adds 32 bytes to the memory size during initialization so it can access beyond end of memory when checking fields. See the code there.

If I recall correctly, the statement at line 2598 was put in to avoid that. At the time the line was added, it was at the end of IS596.MLC. Now other data follows it, so the line can be deleted. If it's okay with you, I will just do this as part of the current PR. Okay?

File: rt/mlc/IS596.MLC (hunk 2 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC:477]: StrBeg is used in the TRTR tests. The TRT tests only use StrNone, StrMid, and StrEnd. The TRTR tests only use StrNone, StrMid, and StrBeg. Don't see a need for having a test for the "first" byte to be a space (first = leftmost byte for TRT, rightmost byte for TRTR). If you feel strongly about this, I will add these tests. Let me know.

File: rt/mlc/IS596.MLC (hunk 3 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC:859]: See comment I added for your TRT comment.

File: rt/mlc/IS596.MLC (hunk 4 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC:911]: If you look at the #596 changes for the TRTR instruction, you will see that the setting of GR1 is done only if there is a space in the source string. The "31-bit only" tests were created specifically for the case a space is anywhere in the string, so using StrNone would not modify GR1. That was tested in the previous TRTR test set (GR1 bit 32 = 1). Don't see a need to do it again. As for not using StrEnd, see above TRT comments.

File: rt/mlc/IS596.MLC (hunk 1 above)
Comment (abekornelis) [reply rt/mlc/IS596.MLC]: that's fine with me.

File: rt/mlc/IS596.MLC (hunk 2 above)
Comment (abekornelis) [reply rt/mlc/IS596.MLC:477]: First position is (to my  mind) an edge case. One which I've seen fail (totally different context) in my own work. I would really appreciate if you're okay to add this test case.

File: rt/mlc/IS596.MLC (hunk 3 above)
Comment (abekornelis) [reply rt/mlc/IS596.MLC:859]: Same line of reasoning, same answer.

File: rt/mlc/IS596.MLC (hunk 4 above)
Comment (abekornelis) [reply rt/mlc/IS596.MLC:911]: For StrNone - agreed, no problem.
For StrEnd, again same reasoning, same answer (and same appreciation)

File: rt/mlc/IS596.MLC (hunk 3 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC:859]: Added TRTR tests for StrEnd. Push coming shortly.

File: rt/mlc/IS596.MLC (hunk 1 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC]: Deleted line. Push to clone coming shortly. Will also contain the TRT StrBeg tests and TRTR StrEnd tests for GR1 bit 32 equal to 1.  None for the TRTR GR1 bit 32 equal to 0.

File: rt/mlc/IS596.MLC (hunk 2 above)
Comment (jyganci) [reply rt/mlc/IS596.MLC:477]: TRT tests for StrBeg added. Push coming shortly.

**Review by abekornelis — APPROVED** [review]
Re-tested with latest changes. All okay. Thanks!
